### PR TITLE
piping and cloud files ops; closes #8; closes #19; closes #89

### LINKED ISSUE #8: cloud files
containers:
- [x] create
- [x] get
- [x] update
- [x] delete

objects:
- [x] create
- [x] get
- [x] update
- [x] delete
- [ ] copy
- [x] download

### LINKED ISSUE #19: Piping
We should be able to pipe in and out from this tool. Currently, only out is possible.

Also, we should look into being opinionated about piping from this tool to itself (see here: https://github.com/jrperritt/rack/issues/17#issuecomment-109850039)

### LINKED ISSUE #89: Go Panic when generating keypair
```
/data/src/rack [cloud-files-ops-and-piping] $ ./rack --region iad servers keypair generate --keypair rack-cli
panic: interface conversion: interface is *keypairs.CreateOpts, not *keypaircommands.paramsGenerate

goroutine 1 [running]:
github.com/jrperritt/rack/serverscommands/keypaircommands.(*commandGenerate).HandleSingle(0xc208038048, 0xc2082536d0, 0x0, 0x0)
	/Users/ken/go/src/github.com/jrperritt/rack/serverscommands/keypaircommands/generate.go:80 +0x12d
github.com/jrperritt/rack/handler.Handle(0x1012198, 0xc208038048)
	/Users/ken/go/src/github.com/jrperritt/rack/handler/handle.go:133 +0x10fe
github.com/jrperritt/rack/serverscommands/keypaircommands.actionGenerate(0xc2080882a0)
	/Users/ken/go/src/github.com/jrperritt/rack/serverscommands/keypaircommands/generate.go:50 +0xd4
github.com/codegangsta/cli.Command.Run(0x434390, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc208042180, 0x5d, 0x4464b0, ...)
	/Users/ken/go/src/github.com/codegangsta/cli/command.go:118 +0x1044
github.com/codegangsta/cli.(*App).RunAsSubcommand(0xc208066500, 0xc2080880e0, 0x0, 0x0)
	/Users/ken/go/src/github.com/codegangsta/cli/app.go:257 +0xc2a
github.com/codegangsta/cli.Command.startApp(0x41d310, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x47bbf0, 0x22, 0x0, ...)
	/Users/ken/go/src/github.com/codegangsta/cli/command.go:183 +0x505
github.com/codegangsta/cli.Command.Run(0x41d310, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x47bbf0, 0x22, 0x0, ...)
	/Users/ken/go/src/github.com/codegangsta/cli/command.go:45 +0x14ee
github.com/codegangsta/cli.(*App).RunAsSubcommand(0xc208066400, 0xc208088000, 0x0, 0x0)
	/Users/ken/go/src/github.com/codegangsta/cli/app.go:257 +0xc2a
github.com/codegangsta/cli.Command.startApp(0x420570, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x464e30, 0x1c, 0x0, ...)
	/Users/ken/go/src/github.com/codegangsta/cli/command.go:183 +0x505
github.com/codegangsta/cli.Command.Run(0x420570, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x464e30, 0x1c, 0x0, ...)
	/Users/ken/go/src/github.com/codegangsta/cli/command.go:45 +0x14ee
github.com/codegangsta/cli.(*App).Run(0xc208066300, 0xc20800a000, 0x8, 0x8, 0x0, 0x0)
	/Users/ken/go/src/github.com/codegangsta/cli/app.go:160 +0xd0c
main.main()
	/data/src/rack/main.go:38 +0x310

goroutine 5 [chan receive]:
github.com/jrperritt/rack/handler.func·001()
	/Users/ken/go/src/github.com/jrperritt/rack/handler/context.go:47 +0x6d
created by github.com/jrperritt/rack/handler.(*Context).ListenAndReceive
	/Users/ken/go/src/github.com/jrperritt/rack/handler/context.go:63 +0xcc

goroutine 10 [runnable]:
net/http.(*persistConn).readLoop(0xc2080760b0)
	/usr/local/go/src/net/http/transport.go:928 +0x9ce
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:660 +0xc9f

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 11 [select]:
net/http.(*persistConn).writeLoop(0xc2080760b0)
	/usr/local/go/src/net/http/transport.go:945 +0x41d
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:661 +0xcbc
```

### PR DESCRIPTION
**Pull request by jrperritt**

### CODE REVIEW

**Comment by jrperritt** [comment]
This PR implements Cloud Files operations and allows for piping to STDIN. The design has been changed such that all the logic flows through a single function (`Handle`, located in a new package named `handler`) that accepts an `interface{}` (of type `Commander`) as a parameter (all commands implement the interface). This reduces a lot of the copy-paste between commands.

In addition to the `Commander` `interface{}`, there is also a `PipeHandler` `interface{}` which has 2 additional methods (`StdinField` and `HandlePipe`) and is only implemented by commands that can accept input from STDIN.

Two new `struct` types have been introduced in the same `handler` package: `handler.Resource` and `handler.Context`. `handler.Resource` stores information while processing a single command's HTTP request. `handler.Context` stores global information such as the `cli.Context` for accessing flags, the `gophercloud.ServiceClient` for authentication, and the channel for sending individual results when they're ready.

Some of the things you can do now:
Bulk delete servers: `(echo 5e862de6-ed99-4db5-af58-41e6ddbda125 && echo 1df5fefc-087d-4d49-ac53-188c31f48be4) | rack servers instance delete --stdin id`

Bulk create servers: `(echo rackBuild4 && echo rackBuild5) | rack servers instance create --image-id 587ee022-8dfe-43e8-ad8b-b4efb1fc1d71 --flavor-id 4 --stdin instance`

Note that, at least for now, piping in values to commands will run all the commands identically accept for the value of the piped field. So in the above example, all the servers would have to have the same `flavor-id` and `image-id` parameters.

Bulk generate keypairs: `(echo kp1 && echo kp2) | rack servers keypair generate --stdin keypair`

Stream file download: `rack files object download --container cloudsync --object README.md > out.txt`

Stream file upload: 
`echo out.txt | rack files object upload --container rackPipeTest --object out3.txt --stdin file` 
or 
`rack files object upload --container rackPipeTest --object out3.txt --file out.txt`

Using the `-h` flag after any command will show you if and which field you can pipe to it.

Inconsistencies likely exist, and bugs are expected. Please comment and/or file issues for them.

File: handler/handle.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+		ctx.ErrExit1(resource)
+	}
+
+	if pipeableCommand, ok := command.(PipeHandler); ok && ctx.CLIContext.IsSet("stdin") {
+		stdinField := ctx.CLIContext.String("stdin")
+		if stdinField == pipeableCommand.StdinField() {
+			scanner := bufio.NewScanner(os.Stdin)
+			for scanner.Scan() {
+				item := scanner.Text()
+				localResource := &Resource{}
```
Comment (jrperritt) [inline handler/handle.go]: This line and the 6 below it are shared with the single-resource logic and will likely be refactored in the future to reduce code duplication.

**Comment by kenperkins** [comment]
I found some unexpected behavior when using `cat` vs `echo` piping a file in:

#### echo
```bash
/data/src/rack [cloud-files-ops-and-piping] $ echo globalflags.go | ./rack --region iad files object upload --container kptest2 --object globalflags.go --stdin file
Successfully uploaded object [globalflags.go] to container [kptest2]
```

#### cat

```bash
/data/src/rack [cloud-files-ops-and-piping] $ cat globalflags.go | ./rack --region iad files object upload --container kptest2 --object globalflags.go --stdin file
PROPERTYVALUE
error	open package main: no such file or directory
PROPERTYVALUE
error	open import (: no such file or directory
PROPERTYVALUE
error	open 	"fmt": no such file or directory
PROPERTYVALUE
error	open 	"github.com/codegangsta/cli": no such file or directory
PROPERTYVALUE
error	open ): no such file or directory
PROPERTYVALUE
error	open // outputFlags are global flags (i.e. flags that all commands can use) that let: no such file or directory
PROPERTYVALUE
error	open // users specify the format of the output from a command.: no such file or directory
PROPERTYVALUE
error	open func outputFlags() []cli.Flag {: no such file or directory
PROPERTYVALUE
error	open 	return []cli.Flag{: no such file or directory
PROPERTYVALUE
error	open 	cli.BoolFlag{: no such file or directory
PROPERTYVALUE
error	open 	Name:  "json",: no such file or directory
PROPERTYVALUE
error	open 	Usage: "Return output in JSON format.",: no such file or directory
PROPERTYVALUE
error	open 	},: no such file or directory
PROPERTYVALUE
error	open 	cli.BoolFlag{: no such file or directory
PROPERTYVALUE
error	open 	Name:  "table",: no such file or directory
PROPERTYVALUE
error	open 	Usage: "Return output in tabular format. This is the default output format.",: no such file or directory
PROPERTYVALUE
error	open 	},: no such file or directory
PROPERTYVALUE
error	open 	cli.BoolFlag{: no such file or directory
PROPERTYVALUE
error	open 	Name:  "csv",: no such file or directory
PROPERTYVALUE
error	open 	Usage: "Return output in csv format.",: no such file or directory
PROPERTYVALUE
error	open 	},: no such file or directory
PROPERTYVALUE
error	open 	}: no such file or directory
PROPERTYVALUE
error	open }: no such file or directory
PROPERTYVALUE
error	open // authFlags are global flags (i.e. flags that all commands can use) that let: no such file or directory
PROPERTYVALUE
error	open // users specify authentication parameters.: no such file or directory
PROPERTYVALUE
error	open func authFlags() []cli.Flag {: no such file or directory
PROPERTYVALUE
error	open 	return []cli.Flag{: no such file or directory
PROPERTYVALUE
error	open 	cli.StringFlag{: no such file or directory
PROPERTYVALUE
error	open 	Name:  "username",: no such file or directory
PROPERTYVALUE
error	open 	Usage: "The username with which to authenticate.",: no such file or directory
PROPERTYVALUE
error	open 	},: no such file or directory
PROPERTYVALUE
error	open 	cli.StringFlag{: no such file or directory
PROPERTYVALUE
error	open 	Name:  "apikey",: no such file or directory
PROPERTYVALUE
error	open 	Usage: "The API key with which to authenticate.",: no such file or directory
PROPERTYVALUE
error	open 	},: no such file or directory
PROPERTYVALUE
error	open 	cli.StringFlag{: no such file or directory
PROPERTYVALUE
error	open 	Name:  "authurl",: no such file or directory
PROPERTYVALUE
error	open 	Usage: "The endpoint to which authenticate.",: no such file or directory
PROPERTYVALUE
error	open 	},: no such file or directory
PROPERTYVALUE
error	open 	cli.StringFlag{: no such file or directory
PROPERTYVALUE
error	open 	Name:  "region",: no such file or directory
PROPERTYVALUE
error	open 	Usage: "The region to which authenticate.",: no such file or directory
PROPERTYVALUE
error	open 	},: no such file or directory
PROPERTYVALUE
error	open 	cli.StringFlag{: no such file or directory
PROPERTYVALUE
error	open 	Name:  "profile",: no such file or directory
PROPERTYVALUE
error	open 	Usage: "The config file profile to use for authentication.",: no such file or directory
PROPERTYVALUE
error	open 	},: no such file or directory
PROPERTYVALUE
error	open 	cli.BoolFlag{: no such file or directory
PROPERTYVALUE
error	open 	Name:  "no-cache",: no such file or directory
PROPERTYVALUE
error	open 	Usage: "Don't get or set authentication credentials in the rack cache.",: no such file or directory
PROPERTYVALUE
error	open 	},: no such file or directory
PROPERTYVALUE
error	open 	}: no such file or directory
PROPERTYVALUE
error	open }: no such file or directory
PROPERTYVALUE
error	open // globalFlags returns the flags that can be used after `rack` in a command, such as: no such file or directory
PROPERTYVALUE
error	open // output flags and authentication flags.: no such file or directory
PROPERTYVALUE
error	open func globalFlags() []cli.Flag {: no such file or directory
PROPERTYVALUE
error	open 	gFlags := outputFlags(): no such file or directory
PROPERTYVALUE
error	open 	gFlags = append(gFlags, authFlags()...): no such file or directory
PROPERTYVALUE
error	open 	return gFlags: no such file or directory
PROPERTYVALUE
error	open }: no such file or directory
PROPERTYVALUE
error	open // completeGlobals returns the options for completing global flags.: no such file or directory
PROPERTYVALUE
error	open func completeGlobals(vals []interface{}) {: no such file or directory
PROPERTYVALUE
error	open 	for _, val := range vals {: no such file or directory
PROPERTYVALUE
error	open 	switch val.(type) {: no such file or directory
PROPERTYVALUE
error	open 	case cli.StringFlag:: no such file or directory
PROPERTYVALUE
error	open 	fmt.Println("--" + val.(cli.StringFlag).Name): no such file or directory
PROPERTYVALUE
error	open 	case cli.IntFlag:: no such file or directory
PROPERTYVALUE
error	open 	fmt.Println("--" + val.(cli.IntFlag).Name): no such file or directory
PROPERTYVALUE
error	open 	case cli.BoolFlag:: no such file or directory
PROPERTYVALUE
error	open 	fmt.Println("--" + val.(cli.BoolFlag).Name): no such file or directory
PROPERTYVALUE
error	open 	case cli.Command:: no such file or directory
PROPERTYVALUE
error	open 	fmt.Println(val.(cli.Command).Name): no such file or directory
PROPERTYVALUE
error	open 	default:: no such file or directory
PROPERTYVALUE
error	open 	continue: no such file or directory
PROPERTYVALUE
error	open 	}: no such file or directory
PROPERTYVALUE
error	open 	}: no such file or directory
PROPERTYVALUE
error	open }: no such file or directory
PROPERTYVALUE
error	open // globalOptions returns the options (flags and commands) that can be used after: no such file or directory
PROPERTYVALUE
error	open // `rack` in a command. For example, `rack --json`, `rack servers`, and: no such file or directory
PROPERTYVALUE
error	open // `rack --json servers` are all legitimate command prefixes.: no such file or directory
PROPERTYVALUE
error	open func globalOptions(app *cli.App) []interface{} {: no such file or directory
PROPERTYVALUE
error	open 	var i []interface{}: no such file or directory
PROPERTYVALUE
error	open 	globalFlags := globalFlags(): no such file or directory
PROPERTYVALUE
error	open 	for _, globalFlag := range globalFlags {: no such file or directory
PROPERTYVALUE
error	open 	i = append(i, globalFlag): no such file or directory
PROPERTYVALUE
error	open 	}: no such file or directory
PROPERTYVALUE
error	open 	for _, cmd := range app.Commands {: no such file or directory
PROPERTYVALUE
error	open 	i = append(i, cmd): no such file or directory
PROPERTYVALUE
error	open 	}: no such file or directory
PROPERTYVALUE
error	open 	return i: no such file or directory
PROPERTYVALUE
error	open }: no such file or directory
Successfully uploaded object [globalflags.go] to container [kptest2]
Successfully uploaded object [globalflags.go] to container [kptest2]
Successfully uploaded object [globalflags.go] to container [kptest2]
Successfully uploaded object [globalflags.go] to container [kptest2]
Successfully uploaded object [globalflags.go] to container [kptest2]
Successfully uploaded object [globalflags.go] to container [kptest2]
Successfully uploaded object [globalflags.go] to container [kptest2]
Successfully uploaded object [globalflags.go] to container [kptest2]
```

**Comment by jrperritt** [comment]
`echo` sends the file name. Only `file` is supported for `stdin` currently. `cat` would be for `content`.

**Comment by kenperkins** [comment]
Interesting, so there's a different command to cat the contents vs the file name itself. Thanks for the clarification!

**Comment by kenperkins** [comment]
So it appears `content` is not supported yet, by design?

**Comment by jrperritt** [comment]
Yes.
```
jrp$ rack files object upload -h
NAME:
   upload - rack [GLOBALS] files object upload --container <containerName> --object <objectName> [OPTIONS]

USAGE:
   command upload [command options] [arguments...]

DESCRIPTION:
   Uploads an object

OPTIONS:
   --container 	[required] The name of the container to upload the object upload
   --object 	[required] The name the object should have in the Cloud Files container
   --content 	[optional; required if `file` or `stdin` isn't provided] The string contents to upload
   --file 	[optional; required if `content` or `stdin` isn't provided] The file name containing the contents to upload
   --stdin 	[optional; required if `file` or `content` isn't provided] The field being piped to STDIN, if any. Valid values are: file
```

**Comment by kenperkins** [comment]
@jrperritt yea, that's what I saw. Was just making sure I wasn't _doing it wrong_. That said this PR makes me :grinning:

**Comment by jrperritt** [comment]
:shipit:

**Pull request merged**